[BUG] Inconsistent data_view.namespaces handling in Terraform Provider elasticstack v0.11.13

### **Describe the bug**

When upgrading the Terraform provider `elastic/elasticstack` from version `0.11.8` to `0.11.13`, the `terraform apply` command fails due to inconsistencies in the `elasticstack_kibana_data_view` resource. The error indicates an unexpected change in the `data_view.namespaces` attribute, causing Terraform to detect an inconsistent state.

----------

### **To Reproduce**

Steps to reproduce the behavior:

1.  Use the following Terraform configuration:
    
    ```hcl
    terraform {
      required_providers {
        elasticstack = {
          source  = "elastic/elasticstack"
          version = "0.11.8"
        }
      }
    }
    
    [...]
    
    resource "elasticstack_kibana_data_view" "my_data_view" {
      data_view = {
        name            = "logsaaaaa-*"
        title           = "elas*"
        time_field_name = "@timestamp"
        namespaces      = ["Prod","NonProd"]
      }
    }
    
    ```
    


2.  Run `terraform plan` and  the `terraform apply`

3. Change `version = "0.11.8"` to `version = "0.11.13"`

4.  Run `terraform init -upgrade`
   
5.  Run `terraform plan` and  the `terraform apply` ( with provider version `0.11.13`.)
    
6.  Observe the error message:
    
    ```
│ Error: Provider produced inconsistent result after apply
│
│ When applying changes to elasticstack_kibana_data_view.my_data_view, provider "provider[\"registry.terraform.io/elastic/elasticstack\"]" produced an
│ unexpected new value: .data_view.namespaces[0]: was cty.StringVal("Prod"), but now cty.StringVal("default").
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
    ```
    
7.  The issue is related to the `data_view.namespaces` field,  even if the value didn't change. Terraform apply  unexpectedly fails.
    

----------

### **Expected behavior**

Terraform should correctly apply without changes .

----------

### **Debug output**
```
2025-02-07T11:05:51.970+0100 [TRACE] provider.terraform-provider-elasticstack_v0.11.13.exe: Received downstream response:

tf_resource_type=elasticstack_kibana_data_view tf_rpc=ApplyResourceChange

@caller=github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/internal/tf6serverlogging/downstream_request.go:42 @module=sdk.proto

diagnostic_error_count=0 diagnostic_warning_count=0 tf_proto_version=6.7 tf_provider_addr=registry.terraform.io/elastic/elasticstack tf_req_duration_ms=1067

tf_req_id=75f0b256-0cf5-451f-01e1-8d299ebfe02c timestamp="2025-02-07T11:05:51.970+0100"

2025-02-07T11:05:51.970+0100 [TRACE] provider.terraform-provider-elasticstack_v0.11.13.exe: Served request:

@caller=github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov6/tf6server/server.go:878 tf_resource_type=elasticstack_kibana_data_view

tf_rpc=ApplyResourceChange tf_provider_addr=registry.terraform.io/elastic/elasticstack tf_req_id=75f0b256-0cf5-451f-01e1-8d299ebfe02c @module=sdk.proto

tf_proto_version=6.7 timestamp="2025-02-07T11:05:51.970+0100"

2025-02-07T11:05:51.971+0100 [TRACE] maybeTainted: elasticstack_kibana_data_view.my_data_view encountered an error during creation, so it is now marked as

tainted

2025-02-07T11:05:51.971+0100 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/elastic/elasticstack" is in the global cache

2025-02-07T11:05:51.971+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for elasticstack_kibana_data_view.my_data_view

2025-02-07T11:05:51.971+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for elasticstack_kibana_data_view.my_data_view

2025-02-07T11:05:51.971+0100 [TRACE] evalApplyProvisioners: elasticstack_kibana_data_view.my_data_view is tainted, so skipping provisioning

2025-02-07T11:05:51.971+0100 [TRACE] maybeTainted: elasticstack_kibana_data_view.my_data_view was already tainted, so nothing to do

2025-02-07T11:05:51.971+0100 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/elastic/elasticstack" is in the global cache

2025-02-07T11:05:51.971+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for elasticstack_kibana_data_view.my_data_view

2025-02-07T11:05:51.971+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for elasticstack_kibana_data_view.my_data_view

2025-02-07T11:05:51.971+0100 [TRACE] statemgr.Filesystem: have already backed up original terraform.tfstate to terraform.tfstate.backup on a previous write

2025-02-07T11:05:51.982+0100 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 83

2025-02-07T11:05:51.982+0100 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate

2025-02-07T11:05:52.006+0100 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot

2025-02-07T11:05:52.006+0100 [ERROR] vertex "elasticstack_kibana_data_view.my_data_view" error: Provider produced inconsistent result after apply

2025-02-07T11:05:52.006+0100 [ERROR] vertex "elasticstack_kibana_data_view.my_data_view" error: Provider produced inconsistent result after apply

2025-02-07T11:05:52.006+0100 [TRACE] vertex "elasticstack_kibana_data_view.my_data_view": visit complete, with errors

2025-02-07T11:05:52.006+0100 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/elastic/elasticstack\"] (close)" errored, so skipping

2025-02-07T11:05:52.006+0100 [TRACE] dag/walk: upstream of "root" errored, so skipping

2025-02-07T11:05:52.006+0100 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/elastic/elasticstack" is in the global cache

2025-02-07T11:05:52.006+0100 [TRACE] terraform.contextPlugins: Schema for provider "registry.terraform.io/elastic/ec" is in the global cache

2025-02-07T11:05:52.006+0100 [TRACE] statemgr.Filesystem: have already backed up original terraform.tfstate to terraform.tfstate.backup on a previous write

2025-02-07T11:05:52.018+0100 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 84

2025-02-07T11:05:52.018+0100 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate

[31mÔòÀ[0m[0m

[31mÔöé[0m [0m[1m[31mError: [0m[0m[1mProvider produced inconsistent result after apply[0m

[31mÔöé[0m [0m

[31mÔöé[0m [0m[0mWhen applying changes to elasticstack_kibana_data_view.my_data_view,

[31mÔöé[0m [0mprovider "provider[\"registry.terraform.io/elastic/elasticstack\"]"

[31mÔöé[0m [0mproduced an unexpected new value: .data_view.namespaces[0]: was

[31mÔöé[0m [0mcty.StringVal("Prod"), but now cty.StringVal("default").

[31mÔöé[0m [0m

[31mÔöé[0m [0mThis is a bug in the provider, which should be reported in the provider's

[31mÔöé[0m [0mown issue tracker.

[31mÔòÁ[0m[0m

[31mÔòÀ[0m[0m

[31mÔöé[0m [0m[1m[31mError: [0m[0m[1mProvider produced inconsistent result after apply[0m

[31mÔöé[0m [0m

[31mÔöé[0m [0m[0mWhen applying changes to elasticstack_kibana_data_view.my_data_view,

[31mÔöé[0m [0mprovider "provider[\"registry.terraform.io/elastic/elasticstack\"]"

[31mÔöé[0m [0mproduced an unexpected new value: .data_view.namespaces: new element 2 has

[31mÔöé[0m [0mappeared.

[31mÔöé[0m [0m

[31mÔöé[0m [0mThis is a bug in the provider, which should be reported in the provider's

[31mÔöé[0m [0mown issue tracker.

[31mÔòÁ[0m[0m

2025-02-07T11:05:52.024+0100 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info

2025-02-07T11:05:52.025+0100 [TRACE] statemgr.Filesystem: unlocked by closing terraform.tfstate

2025-02-07T11:05:52.026+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"

2025-02-07T11:05:52.033+0100 [INFO] provider: plugin process exited:

plugin=.terraform/providers/registry.terraform.io/elastic/elasticstack/0.11.13/windows_amd64/terraform-provider-elasticstack_v0.11.13.exe id=25720

2025-02-07T11:05:52.033+0100 [DEBUG] provider: plugin exited
````
----------

### **Versions (please complete the following information):**

-   **OS:** Windows 11
-   **Terraform Version:** 1.9.6
-   **Provider Version:**
    -   Working: `0.11.8`
    -   Failing: `0.11.13`
-   **Elasticsearch Version:** 8.17.1 (Elastic Cloud)

----------

### **Additional context**

-   The error only occurs when `data_view.namespaces` contains multiple values (e.g., `["Prod","NonProd"]`).
-   The same Terraform configuration works fine with provider version `0.11.8`.
-   No changes were made to the Terraform code except for updating the provider version.
-   The issue seems to be a regression in the provider when handling `namespaces` in `elasticstack_kibana_data_view`.
- If namespaces is commented apply works fine: 
```hcl
resource  "elasticstack_kibana_data_view"  "my_data_view" {

data_view  =  {
name  =  "logsaaaaa-*"
title  =  "elas*"
time_field_name  =  "@timestamp"
#namespaces = ["Prod","NonProd"]
}
}

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 - Fix a provider crash when interacting with elasticstack_kibana_data_view resources created with 0.11.0. ([#979](https://github.com/elastic/terraform-provider-elasticstack/pull/979))
 - Add `max_primary_shard_docs` condition to ILM rollover ([#845](https://github.com/elastic/terraform-provider-elasticstack/pull/845))
 - Add missing entries to `data_view.field_formats.params` ([#1001](https://github.com/elastic/terraform-provider-elasticstack/pull/1001))
+- Fix namespaces inconsistency when creating elasticstack_kibana_data_view resources ([#1234](https://github.com/elastic/terraform-provider-elasticstack/pull/1234))
 
 ## [0.11.13] - 2025-01-09
 
@@ -19,6 +20,7 @@
 - Support multiple group by fields in SLOs ([#870](https://github.com/elastic/terraform-provider-elasticstack/pull/878)). This changes to type of the `group_by` attribute of the `elasticstack_kibana_slo` resource from a String to a list of Strings. Any existing SLO defintions will need to update `group_by = "field"` to `group_by = ["field"]`.
 
 ### Changes
+
 - Handle NPE in integration policy secrets ([#946](https://github.com/elastic/terraform-provider-elasticstack/pull/946))
 - Use the auto-generated OAS schema from elastic/kibana for the Fleet API. ([#834](https://github.com/elastic/terraform-provider-elasticstack/issues/834))
 - Support description in `elasticstack_elasticsearch_security_role` data sources. ([#884](https://github.com/elastic/terraform-provider-elasticstack/pull/884))

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,7 +3,7 @@
 - Fix a provider crash when interacting with elasticstack_kibana_data_view resources created with 0.11.0. ([#979](https://github.com/elastic/terraform-provider-elasticstack/pull/979))
 - Add `max_primary_shard_docs` condition to ILM rollover ([#845](https://github.com/elastic/terraform-provider-elasticstack/pull/845))
 - Add missing entries to `data_view.field_formats.params` ([#1001](https://github.com/elastic/terraform-provider-elasticstack/pull/1001))
-- Fix namespaces inconsistency when creating elasticstack_kibana_data_view resources ([#1234](https://github.com/elastic/terraform-provider-elasticstack/pull/1234))
+- Fix namespaces inconsistency when creating elasticstack_kibana_data_view resources ([#1011](https://github.com/elastic/terraform-provider-elasticstack/pull/1011))
 
 ## [0.11.13] - 2025-01-09
 

--- a/internal/kibana/data_view/create.go
+++ b/internal/kibana/data_view/create.go
@@ -35,7 +35,7 @@ func (r *DataViewResource) Create(ctx context.Context, req resource.CreateReques
 		return
 	}
 
-	diags = planModel.populateFromAPI(ctx, dataView)
+	diags = planModel.populateFromAPI(ctx, dataView, spaceID)
 	resp.Diagnostics.Append(diags...)
 	if resp.Diagnostics.HasError() {
 		return

--- a/internal/kibana/data_view/models.go
+++ b/internal/kibana/data_view/models.go
@@ -12,7 +12,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
-func (model *dataViewModel) populateFromAPI(ctx context.Context, data *kbapi.DataViewsDataViewResponseObject) diag.Diagnostics {
+func (model *dataViewModel) populateFromAPI(ctx context.Context, data *kbapi.DataViewsDataViewResponseObject, spaceID string) diag.Diagnostics {
 	if data == nil {
 		return nil
 	}
@@ -60,10 +60,15 @@ func (model *dataViewModel) populateFromAPI(ctx context.Context, data *kbapi.Dat
 
 		// Keep the original ordering if equal but unordered
 		// The API response is ordered by name.
-		if len(existing) == len(incoming) {
+		// Additionally, allow for the response containing an extra namespace that is the current SpaceID
+		// If the SpaceID is not included in the `namespaces` field, when trying to GET the object it will 404
+		if (len(existing) == len(incoming)) || (len(existing) == len(incoming)-1) {
 			useExisting := true
 			for _, ns := range existing {
 				if !slices.Contains(incoming, ns) {
+					if ns == spaceID {
+						continue
+					}
 					useExisting = false
 					break
 				}

--- a/internal/kibana/data_view/models_test.go
+++ b/internal/kibana/data_view/models_test.go
@@ -100,13 +100,115 @@ func TestPopulateFromAPI(t *testing.T) {
 				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
 			},
 		},
+		{
+			// When sending no value, the response from Kibana is ["default"]
+			name: "handleNamespaces_null_default",
+			existingModel: dataViewModel{
+				ID:      types.StringValue("default/id"),
+				SpaceID: types.StringValue("default"),
+				DataView: utils.ObjectValueFrom(ctx, &innerModel{
+					ID:              types.StringValue("id"),
+					Namespaces:      utils.ListValueFrom[string](ctx, nil, types.StringType, path.Root("data_view").AtName("namespaces"), &diags),
+					SourceFilters:   types.ListNull(types.StringType),
+					FieldAttributes: types.MapNull(getFieldAttrElemType()),
+					RuntimeFieldMap: types.MapNull(getRuntimeFieldMapElemType()),
+					FieldFormats:    types.MapNull(getFieldFormatElemType()),
+				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
+			},
+			response: kbapi.DataViewsDataViewResponseObject{
+				DataView: &kbapi.DataViewsDataViewResponseObjectInner{
+					Id:         utils.Pointer("id"),
+					Namespaces: &[]string{"default"},
+				},
+			},
+			expectedModel: dataViewModel{
+				ID:      types.StringValue("default/id"),
+				SpaceID: types.StringValue("default"),
+				DataView: utils.ObjectValueFrom(ctx, &innerModel{
+					ID:              types.StringValue("id"),
+					Namespaces:      utils.ListValueFrom[string](ctx, nil, types.StringType, path.Root("data_view").AtName("namespaces"), &diags),
+					SourceFilters:   types.ListNull(types.StringType),
+					FieldAttributes: types.MapNull(getFieldAttrElemType()),
+					RuntimeFieldMap: types.MapNull(getRuntimeFieldMapElemType()),
+					FieldFormats:    types.MapNull(getFieldFormatElemType()),
+				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
+			},
+		},
+		{
+			// When sending the SpaceID as the namespace, the response from Kibana should be the same
+			name: "handleNamespaces_populated_default",
+			existingModel: dataViewModel{
+				ID:      types.StringValue("space_id/dataview_id"),
+				SpaceID: types.StringValue("space_id"),
+				DataView: utils.ObjectValueFrom(ctx, &innerModel{
+					ID:              types.StringValue("dataview_id"),
+					Namespaces:      utils.ListValueFrom(ctx, []string{"space_id"}, types.StringType, path.Root("data_view").AtName("namespaces"), &diags),
+					SourceFilters:   types.ListNull(types.StringType),
+					FieldAttributes: types.MapNull(getFieldAttrElemType()),
+					RuntimeFieldMap: types.MapNull(getRuntimeFieldMapElemType()),
+					FieldFormats:    types.MapNull(getFieldFormatElemType()),
+				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
+			},
+			response: kbapi.DataViewsDataViewResponseObject{
+				DataView: &kbapi.DataViewsDataViewResponseObjectInner{
+					Id:         utils.Pointer("dataview_id"),
+					Namespaces: &[]string{"space_id"},
+				},
+			},
+			expectedModel: dataViewModel{
+				ID:      types.StringValue("space_id/dataview_id"),
+				SpaceID: types.StringValue("space_id"),
+				DataView: utils.ObjectValueFrom(ctx, &innerModel{
+					ID:              types.StringValue("dataview_id"),
+					Namespaces:      utils.ListValueFrom(ctx, []string{"space_id"}, types.StringType, path.Root("data_view").AtName("namespaces"), &diags),
+					SourceFilters:   types.ListNull(types.StringType),
+					FieldAttributes: types.MapNull(getFieldAttrElemType()),
+					RuntimeFieldMap: types.MapNull(getRuntimeFieldMapElemType()),
+					FieldFormats:    types.MapNull(getFieldFormatElemType()),
+				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
+			},
+		},
+		{
+			// When sending a populated list, the response from Kibana should be the same list
+			name: "handleNamespaces_populated_default",
+			existingModel: dataViewModel{
+				ID:      types.StringValue("test/placeholder"),
+				SpaceID: types.StringValue("test"),
+				DataView: utils.ObjectValueFrom(ctx, &innerModel{
+					ID:              types.StringValue("placeholder"),
+					Namespaces:      utils.ListValueFrom(ctx, []string{"ns1", "ns2"}, types.StringType, path.Root("data_view").AtName("namespaces"), &diags),
+					SourceFilters:   types.ListNull(types.StringType),
+					FieldAttributes: types.MapNull(getFieldAttrElemType()),
+					RuntimeFieldMap: types.MapNull(getRuntimeFieldMapElemType()),
+					FieldFormats:    types.MapNull(getFieldFormatElemType()),
+				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
+			},
+			response: kbapi.DataViewsDataViewResponseObject{
+				DataView: &kbapi.DataViewsDataViewResponseObjectInner{
+					Id:         utils.Pointer("placeholder"),
+					Namespaces: &[]string{"ns1", "ns2"},
+				},
+			},
+			expectedModel: dataViewModel{
+				ID:      types.StringValue("test/placeholder"),
+				SpaceID: types.StringValue("test"),
+				DataView: utils.ObjectValueFrom(ctx, &innerModel{
+					ID:              types.StringValue("placeholder"),
+					Namespaces:      utils.ListValueFrom(ctx, []string{"ns1", "ns2"}, types.StringType, path.Root("data_view").AtName("namespaces"), &diags),
+					SourceFilters:   types.ListNull(types.StringType),
+					FieldAttributes: types.MapNull(getFieldAttrElemType()),
+					RuntimeFieldMap: types.MapNull(getRuntimeFieldMapElemType()),
+					FieldFormats:    types.MapNull(getFieldFormatElemType()),
+				}, getDataViewAttrTypes(), path.Root("data_view"), &diags),
+			},
+		},
 	}
 
 	require.Empty(t, diags)
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			diags := tt.existingModel.populateFromAPI(ctx, &tt.response)
+			diags := tt.existingModel.populateFromAPI(ctx, &tt.response, tt.existingModel.SpaceID.ValueString())
 
 			require.Equal(t, tt.expectedModel, tt.existingModel)
 			require.Empty(t, diags)

--- a/internal/kibana/data_view/models_test.go
+++ b/internal/kibana/data_view/models_test.go
@@ -186,7 +186,7 @@ func TestPopulateFromAPI(t *testing.T) {
 			response: kbapi.DataViewsDataViewResponseObject{
 				DataView: &kbapi.DataViewsDataViewResponseObjectInner{
 					Id:         utils.Pointer("placeholder"),
-					Namespaces: &[]string{"ns1", "ns2"},
+					Namespaces: &[]string{"test", "ns1", "ns2"},
 				},
 			},
 			expectedModel: dataViewModel{

--- a/internal/kibana/data_view/read.go
+++ b/internal/kibana/data_view/read.go
@@ -34,7 +34,7 @@ func (r *DataViewResource) Read(ctx context.Context, req resource.ReadRequest, r
 		return
 	}
 
-	diags = stateModel.populateFromAPI(ctx, dataView)
+	diags = stateModel.populateFromAPI(ctx, dataView, spaceID)
 	resp.Diagnostics.Append(diags...)
 	if resp.Diagnostics.HasError() {
 		return

--- a/internal/kibana/data_view/update.go
+++ b/internal/kibana/data_view/update.go
@@ -35,7 +35,7 @@ func (r *DataViewResource) Update(ctx context.Context, req resource.UpdateReques
 		return
 	}
 
-	diags = planModel.populateFromAPI(ctx, dataView)
+	diags = planModel.populateFromAPI(ctx, dataView, spaceID)
 	resp.Diagnostics.Append(diags...)
 	if resp.Diagnostics.HasError() {
 		return

